pkg/controllers/csi/metadata/sqlite_gorm_client.go

```diff
@@ -0,0 +1,197 @@
+
+	CreateAppMount(ctx context.Context, appMount *AppMount) error
+	UpdateAppMount(ctx context.Context, appMount *AppMount) error
+	DeleteAppMount(ctx context.Context, appMount *AppMount) error
+	ReadAppMountByVolumeMetaID(ctx context.Context, volumeMetaID string) (*AppMount, error)
+	ReadAppMounts(ctx context.Context) ([]AppMount, error)
+}
+
+type DBConn struct {
+	db *gorm.DB
+}
+
+var _ DBAccess = (*DBConn)(nil)
+
+// NewDBAccess creates a new gorm db connection to the database.
+func NewDBAccess(path string) (DBConn, error) {
+	// we should explicitly enable foreign_keys for sqlite
+	if strings.Contains(path, "?") {
+		path += "&_foreign_keys=on"
+	} else {
+		path += "?_foreign_keys=on"
+	}
+
+	db, err := gorm.Open(sqlite.Open(path), &gorm.Config{Logger: logger.Default})
+
+	if err != nil {
+		return DBConn{}, err
+	}
+
+	return DBConn{db: db}, nil
+}
+
+// SchemaMigration runs gormigrate migrations to create tables
+func (conn *DBConn) SchemaMigration(_ context.Context) error {
+	m := gormigrate.New(conn.db, gormigrate.DefaultOptions, []*gormigrate.Migration{})
+	m.InitSchema(func(tx *gorm.DB) error {
+		err := tx.AutoMigrate(
+			&TenantConfig{},
+			&CodeModule{},
+			&OSMount{},
+			&AppMount{},
+			&VolumeMeta{},
+		)
+		if err != nil {
+			return err
+		}
+		// all other constraints, indexes, etc...
+		return nil
+	})
+
+	_ = m.Migrate()
+
+	return gormigrate.New(conn.db, gormigrate.DefaultOptions, []*gormigrate.Migration{
+		{
+			ID:      "202403041200",
+			Migrate: dataMigration,
+			Rollback: func(tx *gorm.DB) error {
+				return nil
+			},
+		},
+	}).Migrate()
+}
+
+func (conn *DBConn) CreateTenantConfig(ctx context.Context, tenantConfig *TenantConfig) error {
+	return conn.db.WithContext(ctx).Create(tenantConfig).Error
+}
+
+func (conn *DBConn) UpdateTenantConfig(ctx context.Context, tenantConfig *TenantConfig) error {
+	return conn.db.WithContext(ctx).Updates(tenantConfig).Error
+}
+
+func (conn *DBConn) DeleteTenantConfig(ctx context.Context, tenantConfig *TenantConfig) error {
+	return conn.db.WithContext(ctx).Delete(tenantConfig).Error
+}
+
+func (conn *DBConn) CreateCodeModule(ctx context.Context, codeModule *CodeModule) error {
+	return conn.db.WithContext(ctx).Create(codeModule).Error
+}
+
+func (conn *DBConn) ReadCodeModuleByVersion(ctx context.Context, version string) (*CodeModule, error) {
+	var codeModule CodeModule
+
+	result := conn.db.WithContext(ctx).First(&codeModule, "version = ?", version)
+	if result.Error != nil {
+		return nil, result.Error
+	}
+
+	return &codeModule, nil
+}
+
+func (conn *DBConn) DeleteCodeModule(ctx context.Context, codeModule *CodeModule) error {
+	return conn.db.WithContext(ctx).Delete(codeModule).Error
+}
+
+func (conn *DBConn) ReadTenantConfigByTenantUUID(ctx context.Context, uid string) (*TenantConfig, error) {
+	var tenantConfig TenantConfig
+
+	result := conn.db.WithContext(ctx).First(&tenantConfig, "tenant_uuid = ?", uid)
```

Comment:
you can do it in more elegant way IMO:
```
tenantConfig := TenantConfig{TenantUUID: uid}
```
and then:
```
result := conn.db.WithContext(ctx).First(&tenantConfig)
```

Reply:
Interestingly my test failed when called with "" as uid:
```go
_, err = db.ReadTenantConfigByTenantUUID(context.Background(), "")
require.Error(t, err)
``` 
I expect it to fail with an error, but it simply returns the first (and only entry) in the table.

Reply:
true. I tried to make use of our datamodel/framework as much as possible and not duplicate checks, but for some params I was missing.

in this example I would not change the code bc the next check also fails with this code unexpectedly:

```go
_, err = db.ReadTenantConfigByTenantUUID(context.Background(), "unknown")
require.Error(t, err)
```

Reply:
wont change that as it does not work

Reply:
it works if you refactor using this snippet:
```
func (conn *DBConn) ReadTenantConfigByTenantUUID(ctx context.Context, uid string) (*TenantConfig, error) {
	if uid == "" {
		return nil, errors.New("error")
	} // TODO: check will be obsolete with move to sql.NullString

	tenantConfig := &TenantConfig{TenantUUID: uid}

	result := conn.db.WithContext(ctx).Where(tenantConfig).First(tenantConfig)
	if result.Error != nil {
		return nil, result.Error
	}

	return tenantConfig, nil
}
```

Reply:
fixed in 88cad875c715d861a4a61859035b75333f199764